src/blocks/data.js

```diff
@@ -768,6 +777,62 @@ class Scratch3DataBlocks {
 
         this.runtime.requestToolboxExtensionsUpdate();
     }
+
+    serializeVariable (block, target) {
+        const variableName = block.mutation.blockInfo.text;
+        const variable = target.lookupVariableByNameAndType(variableName, Variable.SCALAR_TYPE);
+        const result = [VAR_PRIMITIVE, variableName, variable.id];
+        if (block.topLevel) {
+            result.push(block.x ? Math.round(block.x) : 0);
+            result.push(block.y ? Math.round(block.y) : 0);
+        }
+        return result;
+    }
+
+    deserializeVariable (primitiveArray) {
+        const [primitiveTag, variableName, variableId] = primitiveArray;
+        if (primitiveTag !== VAR_PRIMITIVE) {
+            throw new Error('bad primitive ID on variable!');
+        }
+        // @todo reduce duplication here.
+        // This effectively duplicates some of the content in `getInfo()` as well as some of the logic in
+        // `defineDynamicBlock`. Maybe it should return only the mutation blockInfo, or maybe it should return
+        // something that looks like the block's entry in `getInfo()`. If we do either of those things we'll need to
+        // decide how we want to handle the block's (x,y) position and `topLevel` flag.
+        const result = {
+            opcode: 'variable',
```

Comment:
This opcode is incorrect, the top level block opcode is the "extended" opcode.

Reply:
I was trying to avoid having the extension every mention the extended opcode, since in theory we might sometimes post-process the extension ID so that it isn't what the extension author expects. Instead I fix up the ID after the fact, in the function which calls this one.